float32 remaining_time
